Mathlib/RingTheory/TwoSidedIdeal/Lattice.lean

```diff
@@ -0,0 +1,113 @@
+/-
+Copyright (c) 2024 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+
+import Mathlib.RingTheory.TwoSidedIdeal.Basic
+
+/-!
+# The complete lattice structure on two-sided ideals
+-/
+
+namespace TwoSidedIdeal
+
+variable (R : Type*) [NonUnitalNonAssocRing R]
+
+instance : Sup (TwoSidedIdeal R) where
+  sup := fun I J => { ringCon := I.ringCon ⊔ J.ringCon }
+
+instance : SemilatticeSup (TwoSidedIdeal R) where
+  le_sup_left I J :=  by rw [ringCon_le_iff]; exact le_sup_left
+  le_sup_right I J := by rw [ringCon_le_iff]; exact le_sup_right
+  sup_le I J K h1 h2 := by rw [ringCon_le_iff] at h1 h2 ⊢; exact sup_le h1 h2
+
+section sup
+
+variable {R}
+
+lemma mem_sup_left {I J : TwoSidedIdeal R} {x : R} :
+    x ∈ I → x ∈ I ⊔ J :=
+  fun h => (show I ≤ I ⊔ J from le_sup_left) h
+
+lemma mem_sup_right {I J : TwoSidedIdeal R} {x : R} :
+    x ∈ J → x ∈ I ⊔ J :=
+  fun h => (show J ≤ I ⊔ J from le_sup_right) h
+
+lemma mem_sup {I J : TwoSidedIdeal R} {x : R} :
+    x ∈ I ⊔ J ↔ ∃ y ∈ I, ∃ z ∈ J, y + z = x := by
+  constructor
+  · let s : TwoSidedIdeal R := .mk'
+      {x | ∃ y ∈ I, ∃ z ∈ J, y + z = x}
+      ⟨0, ⟨zero_mem _, ⟨0, ⟨zero_mem _, zero_add _⟩⟩⟩⟩
+      (by rintro _ _ ⟨x, ⟨hx, ⟨y, ⟨hy, rfl⟩⟩⟩⟩ ⟨a, ⟨ha, ⟨b, ⟨hb, rfl⟩⟩⟩⟩;
+          exact ⟨x + a, ⟨add_mem _ hx ha, ⟨y + b, ⟨add_mem _ hy hb, by abel⟩⟩⟩⟩)
+      (by rintro _ ⟨x, ⟨hx, ⟨y, ⟨hy, rfl⟩⟩⟩⟩
+          exact ⟨-x, ⟨neg_mem _ hx, ⟨-y, ⟨neg_mem _ hy, by abel⟩⟩⟩⟩)
+      (by rintro r _ ⟨x, ⟨hx, ⟨y, ⟨hy, rfl⟩⟩⟩⟩
+          exact ⟨_, ⟨mul_mem_left _ _ _ hx, ⟨_, ⟨mul_mem_left _ _ _ hy, mul_add _ _ _ |>.symm⟩⟩⟩⟩)
+      (by rintro r _ ⟨x, ⟨hx, ⟨y, ⟨hy, rfl⟩⟩⟩⟩
+          exact ⟨_, ⟨mul_mem_right _ _ _ hx, ⟨_, ⟨mul_mem_right _ _ _ hy, add_mul _ _ _ |>.symm⟩⟩⟩⟩)
+
+    suffices (I.ringCon ⊔ J.ringCon) ≤ s.ringCon by
+      intro h; convert this h; rw [rel_iff, sub_zero, mem_mk']; rfl
+
+    refine sup_le (fun x y h => ?_) (fun x y h => ?_) <;> rw [rel_iff] at h ⊢ <;> rw [mem_mk']
+    exacts [⟨_, ⟨h, ⟨0, ⟨zero_mem _, add_zero _⟩⟩⟩⟩, ⟨0, ⟨zero_mem _, ⟨_, ⟨h, zero_add _⟩⟩⟩⟩]
+
+  · rintro ⟨y, ⟨hy, ⟨z, ⟨hz, rfl⟩⟩⟩⟩; exact add_mem _ (mem_sup_left hy) (mem_sup_right hz)
+
+end sup
+
+instance : Inf (TwoSidedIdeal R) where
+  inf := fun I J => { ringCon := I.ringCon ⊓ J.ringCon }
```

Comment:
For each of these instances, could you add the obvious `(I ⊓ J).ringCon = I.ringCon ⊓ J.ringCon`-style lemma?

Reply:
done